File: BagOfTokens.go
func bagOfTokensScore(tokens []int, power int) int {
    l := 0
    h := len(tokens) - 1
    score := 0

    sort.Ints(tokens)

    for l <= h {
        if power >= tokens[l] {
            score += 1
            power -= tokens[l]
            l += 1
        } else if l < h && score > 0 {
            score -= 1
            power += tokens[h]
            h -= 1
        } else {
            return score
        }
    }
    return score
}

File: BagOfTokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        int l = 0;
        int h = tokens.size() - 1;
        int score = 0;

        sort(tokens.begin(), tokens.end());

        while(l <= h) {
            if (power >= tokens[l]) {
                score++;
                power -= tokens[l];
                l++;
            } else if (l < h && score > 0) {
                score--;
                power += tokens[h];
                h--;
            } else {
                return score;
            }
        }

        return score;
    }
};